src/steadystate_iterative.jl

```diff
@@ -0,0 +1,147 @@
+using ..timeevolution: nh_hamiltonian, dmaster_h, dmaster_nh, check_master
+
+"""
+    iterative!(rho0, H, J, [method!], args...; [log=false], kwargs...) -> rho[, log]
+
+Compute the steady state density matrix of master equation defined by a
+Hamiltonian and a set of jump operators by solving `L rho = 0` via an iterative
+method provided as argument.
+
+# Arguments
+* `rho0`: Initial density matrix. Note that this gets mutated in-place.
+* `H`: Operator specifying the Hamiltonian.
+* `J`: Vector containing all jump operators.
+* `method!`: The iterative method to be used. Defaults to `IterativeSolvers.bicgstabl!`
+    or `IterativeSolvers.idrs!` depending on arguments' types.
+* `rates=nothing`: Vector or matrix specifying the coefficients (decay rates) for
+    the jump operators. If nothing is specified all rates are assumed to be 1.
+* `Jdagger=dagger.(J)`: Vector containing the hermitian conjugates of the jump
+    operators. If they are not given they are calculated automatically.
+* `args...`: Further arguments are passed on to the iterative solver.
+* `kwargs...`: Further keyword arguments are passed on to the iterative solver.
+See also: [`iterative`](@ref)
+
+Credit for this implementation goes to Z. Denis and F. Vicentini.
+See also https://github.com/Z-Denis/SteadyState.jl
+"""
+function iterative!(rho0::Operator, H::AbstractOperator, J, method! = nothing, args...;
+                    rates=nothing, Jdagger=dagger.(J), kwargs...)
+
+    if method! === nothing
+        if isblascompatible(rho0) && isblascompatible(H) && all(isblascompatible.(J))
+            _method! = IterativeSolvers.bicgstabl!
+        else
+            _method! = IterativeSolvers.idrs!
+        end
+    else
+        _method! = method!
+    end
+
+    # Solution x must satisfy L.x = y with y[1] = tr(x) = 1 and y[j≠1] = 0.
+    M = length(rho0.basis_l)
+    x0 = zeros(eltype(rho0), M^2+1)
+    x0[2:end] .= reshape(rho0.data, M^2)
+
+    y = zeros(eltype(rho0), M^2+1)
+    y[1] = one(y[1])
```

Comment:
Actually, although it should work, it can be made a bit better. Instead of taking just ones, one can take a weight which is e.g. the mean absolute of all non-zero elements in L. This balances how much the trace condition is enforced during the iteration procedure. Also, I am not sure how it is done in here, but I think one should not replace the trace condition by an existing equation but rather add it to the list of equations. This makes the set of equation equation overdetermined but ensures that the otherwise removed line of equation is still there.

Reply:
@PhilipVinc just to make sure, you mean `y[end] = one(eltype(rho0))` here right?

@davidschlegel I don't completely follow. How would the additional equation for the trace condition look?

Reply:
@david-pl yes, sorry, it was a typo.

@davidschlegel This code is already doing that. In the sense that we are solving a system of M^2 + 1 equations. I think what you are talking about is another way to approach this problem but we had seen it has issues sometimes.